src/subtype.c

```diff
@@ -2847,10 +2850,23 @@ static jl_value_t *finish_unionall(jl_value_t *res JL_MAYBE_UNROOTED, jl_varbind
 
     if (res != jl_bottom_type && vb->innervars != NULL) {
         int i;
-        for(i=0; i < jl_array_len(vb->innervars); i++) {
+        for (i = 0; i < jl_array_len(vb->innervars); i++) {
             jl_tvar_t *var = (jl_tvar_t*)jl_array_ptr_ref(vb->innervars, i);
             if (jl_has_typevar(res, var))
                 res = jl_type_unionall((jl_tvar_t*)var, res);
+            // TODO: full dominator analysis for when handling innervars
+            // the `btemp->prev` walk is only giving a sort of post-order guarantee (since we are
+            // iterating 2 trees at once), so once we set `wrap`, there might remain other branches
+            // of the type walk that now may have incomplete bounds: finish those now too
+            jl_varbinding_t *btemp = e->vars;
```

Comment:
Perhaps we still need to make sure there no duplicated innervar in the env (just like the `jl_varbinding_t *wrap = NULL;` trick above)?

And the TODO seems get resolved?

Reply:
And perhaps we'd better not wrap `res` if we update the `innervars`?

Reply:
From my original analysis, I didn't expect that would be possible. There are only 2 sides to the type tree, so either wrapping to the left or wrapping to the right should be needed, but not more than that. With an attempt of asserting for this case, I don't think it is hit either, but the code rearrangement probably makes sense anyways to be more straightforward. I believe a variable would have to switch sides multiple times by combinations to get tricky here, and it doesn't seem like that happens in existing code.